src/easylink/configuration.py

```diff
@@ -45,42 +85,54 @@ class Config:
 
     def __init__(
         self,
-        pipeline_specification: Union[str, Path],
-        input_data: Union[str, Path],
-        computing_environment: Optional[Union[str, Path]],
-        results_dir: Union[str, Path],
+        config_params: Dict,
     ):
-        # Handle pipeline specification
-        self.pipeline = load_yaml(pipeline_specification)
-        self._requires_spark = self._determine_if_spark_is_required(self.pipeline)
-        # Handle input data specification
-        self.input_data = self._load_input_data_paths(input_data)
-        # Handle environment specification
-        self.environment = self._load_computing_environment(computing_environment)
-        # Store path to results directory
-        self.results_dir = Path(results_dir)
-
-        # Extract environment attributes and assign defaults as necessary
-        self.computing_environment = self._get_required_attribute(
-            self.environment, "computing_environment"
-        )
-        self.container_engine = self._get_required_attribute(
-            self.environment, "container_engine"
-        )
-        self.slurm = self.environment.get("slurm", {})  # no defaults for slurm
-        self.implementation_resources = self._get_implementation_resource_requests(
-            self.environment
-        )
-        self.spark = self._get_spark_requests(self.environment, self._requires_spark)
+        super().__init__(layers=["initial_data", "default", "user_configured"])
+        self.update(DEFAULT_ENVIRONMENT, layer="default")
+        self.update(config_params, layer="user_configured")
+        for step in self.pipeline["steps"]:
```

Comment:
this confused me for a second - is there any chance that config_params wouldn't have pipeline specified? I think that's validated up front but don't rememeber for sure

Reply:
It is a key in the config tree. The "pipeline" key itself is set in the configparams:

    return {
        "pipeline": load_yaml(pipeline_specification),
        "input_data": load_input_data_paths(input_data),
        "environment": load_computing_environment(computing_environment),
        "results_dir": Path(results_dir),
    }


independently of the user yaml.
(ln 42)
The pipeline specification itself is a required CLI param, so that's when an error would be thrown if a user doesn't pass in a spec file at all.

Reply:
(copied from comment below)